auction.sympl

```diff
@@ -33,95 +40,324 @@ schema Auction:
-            cvm.error("The bid amount must be superior to initial price id={product_id}")
+            cvm.error(f"The bid amount must be larger than initial price ${auction.initial_price}.")
     else:
         if amount <= last_bid.amount:
-            cvm.error("The bid amount must be superior to previous bid amount id={product_id}")
+            cvm.error(f"The bid amount must be larger than previous bid amount ${last_bid.amount}.")
+
     bid = Bid(bidder = bidder, amount = amount)
     auction.last_bid = bid
+    
     cvm.storage.put(product_id, auction)
-    cvm.job_complete(std.json(auction))
+    if channel == 'PUBLIC':
+        cvm.create_event('Bid Placed', std.json(auction))
+    else:
+        cvm.create_event('Bid Placed', {"channel": channel, "notice": "Bid placed on private auction."})
+
+
+#################################
+# Admin API and Implementations #
+#################################
+
+# Creates the first admin.
+@clientside
+def create_admin(new_admin: KeyAlias) -> None:
+    if check_admin():
+        cvm.error(f'An admin already exists.')
+    else:
+        with PostTxArgs(PUBLIC):
+            _create_admin(new_admin)
+
+@executable
+def _create_admin(new_admin: KeyAlias) -> None:
+    if check_admin():
+        cvm.error(f'An admin already exists.')
+    else:
+        admin = Admin(id=new_admin)
+        cvm.storage.put(Identifier('admin'), admin)
+        cvm.create_event('Admin Created', std.json(admin))
+
+# Updates the admin. This will demote the current admin while creating a new one so only one exists at a time.
+@clientside
+def update_admin(new_admin: KeyAlias) -> None:
+    if check_admin():
+        with PostTxArgs(PUBLIC):
+            _update_admin(new_admin)
+    else:
+        cvm.error('Only the current admin can perform this function.')
+
+@executable
+def _update_admin(new_admin: KeyAlias) -> None:
+    if check_admin():
+        admin = Admin(id=new_admin)
+        cvm.storage.put(Identifier('admin'), admin)
+        cvm.create_event('Admin Updated', std.json(admin))
+    else:
+        cvm.error('Only the current admin can perform this function.')
+
+# Checks if the transaction signer is the admin.
+@helper
+def check_admin() -> bool:
+    existing_admin = cvm.storage.get(PUBLIC, AdminStatic, Identifier('admin'))
+    if isinstance(existing_admin, Admin):
+        admin_ka : str = existing_admin.id
+        tx_ka : str = cvm.tx.key_alias
+        return admin_ka == tx_ka
+    else:
+        return False
+
+
+###########################################
+# Premium Channel API and Implementations #
+###########################################
+
+# Creates the premium channel. No other channels may be created after this one.
+@clientside
+def create_premium_channel() -> None:
+    if check_admin():
+        existing_premium = cvm.storage.get(PUBLIC, PremiumChannelStatic, Identifier('premium'))
+        if isinstance(existing_premium, None):
+            premium = cvm.new_channel('PRM')
```

Comment:
If there's only 1 premium channel, doesn't that mean that you have access to all premium auctions? Not sure that feels very premium :-)

Reply:
Wouldn't a channel per auction be easier to implement?

Reply:
Yeah, that's the idea: there's one channel with elevated privileges (the special, private auctions), hence the comparison to Amazon Prime, Spotify Premium, etc. (each having a paid membership [private channel] for elevated privileges). 

A channel per auction would probably be easier to implement, but since this is for teaching purposes, I didn't want just a copy of Chat (which implements a channel per room).